### PR TITLE
or on NimNode

### PR DESCRIPTION
**Pull request by krux02**
I am using this code, and whenever I don't have it I copy past it. So it might be worth to have it in the macros module.

This in an example code piece where I actually use it (it is not constructed to demostrate the or operator and therefore a bit long):

```nim
proc injectTypes(framebuffer, mesh, arg: NimNode): NimNode {.compileTime.} =
  ## Inject types to the ``render .. do: ...`` node.
  arg.expectKind(nnkDo)
  let formalParams = arg[3]
  formalParams.expectKind(nnkFormalParams)

  let vertexTypeExpr = quote do:
    `mesh`.type.VertexType

  let fragmentTypeExpr = quote do:
    `framebuffer`.type.FragmentType

  let resultType = formalParams[0] or fragmentTypeExpr
  var vertexArg, vertexType, envArg, envType: NimNode

  formalParams.matchAst:
  of nnkFormalParams(_,nnkIdentDefs(`vertexArgX`, `envArgX`,nnkEmpty,_)):
    vertexArg = vertexArgX
    vertexType = vertexTypeExpr
    envArg = envArgX
    envType = nnkVarTy.newTree(bindSym"GlEnvironment")
  of nnkFormalParams(_,nnkIdentDefs(`vertexArgX`, `vertexTypeX`, _), nnkIdentDefs(`envArgX`, `envTypeX`,_)):
    vertexArg = vertexArgX
    vertexType = vertexTypeX or vertexTypeExpr
    envArg = envArgX
    envType = envTypeX or nnkVarTy.newTree(bindSym"GlEnvironment")

  let newParams =
    nnkFormalParams.newTree(
      resultType,
      nnkIdentDefs.newTree(
        vertexArg, vertexType, empty
      ),
      nnkIdentDefs.newTree(
        envArg, envType, empty
      )
    )

  result = arg
  result[3] = newParams
```

The general use case of the ``or`` operator is something I got to know from programming emacs lisp. It could also be implemented for the ``option`` type and for the pointer types such as ``ptr`` and ``ref``.

### CODE REVIEW

**Comment by dom96** [comment]
This looks like a generic "pick the first non-nil value" operation. IIRC other languages have an operator for this. It might be a good idea to adopt this operator and also implement it for nilable types.

**Comment by Araq** [comment]
@dom96 These other languages use ``or`` or ``||``.

File: lib/core/macros.nim [hunk 1]
```diff
@@ -175,6 +175,22 @@ proc `[]=`*(n: NimNode, i: BackwardsIndex, child: NimNode) =
   ## set `n`'s `i`'th child to `child`.
   n[n.len - i.int] = child
 
+template `or`*(a,b: NimNode): NimNode =
```
Comment (Araq) [inline lib/core/macros.nim]: I dunno what is so hard about using the same style the whole rest of the file is written in. In other words, write ``a, b`` with a space after the comma. Just like in written English/German/French/etc.

File: lib/core/macros.nim [hunk 2]
```diff
@@ -175,7 +175,7 @@ proc `[]=`*(n: NimNode, i: BackwardsIndex, child: NimNode) =
   ## set `n`'s `i`'th child to `child`.
   n[n.len - i.int] = child
 
-template `or`*(a,b: NimNode): NimNode =
+template `or`*(x, y: NimNode): NimNode =
   ## Evalutuate ``a`` and when it is not an empty node, return return
```
[the remark this answers is not in the document]
Comment (krux02) [reply lib/core/macros.nim]: Yes that is true. I renamed from ``a,b`` to ``x,y`` because those are the arguments names in all other implementations.

**Comment by krux02** [comment]
I checked why the test was failing. It was not my fault at all. When I compiled it, I did not recompile the compiler, because I was only going to do a library change. But the newest development version breaks this code. So it is not my fault that the test failed.

https://github.com/nim-lang/Nim/issues/8402

**Pull request merged**